File: evolutus-solutions/src/main/java/pl/edu/agh/evolutus/config/EnvironmentConfig.java
package pl.edu.agh.evolutus.config;

import jdk.nashorn.internal.objects.NativeArray;
import jdk.nashorn.internal.runtime.ScriptObject;
import pl.edu.agh.evolutus.environment.BoundaryConditions;
import pl.edu.agh.evolutus.genotype.Genome;
import pl.edu.agh.evolutus.utils.CurrentDirection;
import pl.edu.agh.evolutus.utils.VectorD;
import pl.edu.agh.evolutus.utils.VectorL;

public class EnvironmentConfig {

	private final IConfigJS configJS;

	public EnvironmentConfig(IConfigJS configJS) {
		this.configJS = configJS;
	}

	double unitLengthInMeters() {
		return configJS.unitLengthInMeters();
	}

	double stepDurationInHours() {
		return configJS.stepDurationInHours();
	}

	public VectorL oceanSize() {
		VectorD sizeMeters = VectorD.fromScriptObject(configJS.oceanSize());
		return metersToUnits(sizeMeters);
	}

	public double algaeEnergy() {
		return configJS.algaeEnergy();
	}

	public long initialForamsCount(VectorL positionUnits) {
		VectorD positionMeters = unitsToMeters(positionUnits);
		return configJS.initialForamsCount(positionMeters.x, positionMeters.y, positionMeters.z);
	}

	public double initialEnergy() {
		return configJS.initialEnergy();
	}

	public double initialAlgaeAvailability(VectorL positionUnits) {
		VectorD positionMeters = unitsToMeters(positionUnits);
		return configJS.initialAlgaeAvailability(positionMeters.x, positionMeters.y, positionMeters.z);
	}

	public double algaeGrowth(double insolation) {
		return configJS.algaeGrowth(insolation);
	}

	public double insolation(VectorL positionUnits) {
		VectorD positionMeters = unitsToMeters(positionUnits);
		return configJS.insolation(positionMeters.x, positionMeters.y, positionMeters.z);
	}

	public CurrentDirection currentDirection(VectorL positionUnits) {
		VectorD positionMeters = unitsToMeters(positionUnits);
		ScriptObject currentDirection = configJS.currentDirection(positionMeters.x, positionMeters.y, positionMeters.z);
		return new CurrentDirection(VectorD.fromScriptObject(currentDirection));
	}

	public BoundaryConditions boundaryConditions() {
		return BoundaryConditions.fromString(configJS.boundaryConditions());
	}

	public Genome initialGenome(VectorL positionUnits) {
		VectorD positionMeters = unitsToMeters(positionUnits);
		NativeArray initialGenome = configJS.initialGenome(positionMeters.x, positionMeters.y, positionMeters.z);
		return Genome.fromScriptObject(initialGenome);
	}

	public String crossingOverOperator() {
		return configJS.crossingOverOperator();
	}

	double unitsToMeters(long units) {
		return units * unitLengthInMeters();
	}

	VectorD unitsToMeters(VectorL units) {
		return units.mul(unitLengthInMeters());
	}

	long metersToUnits(double meters) {
		return Math.round(meters / unitLengthInMeters());
	}

	VectorL metersToUnits(VectorD meters) {
		return meters.div(unitLengthInMeters()).toLong();
	}

	double stepsToHours(long steps) {
		return steps * stepDurationInHours();
	}

	long hoursToSteps(double hours) {
		return Math.round(hours / stepDurationInHours());
	}
}

File: evolutus-distribution/src/main/resources/conf/environment_config.js

function oceanSize() {
   return {x: 10, y: 10, z: 1};
}

function algaeEnergy() {
   return 1.0;
}

function initialForamsCount(x, y, z) {
   if (x == 1 && y == 1) {
      return 100;
   } else if (x == 5 && y == 5) {
      return 50;
   }
   return 0;
}

function initialAlgaeAvailability(x, y, z) {
   return 10.0;
}

function algaeGrowth(insolation) {
   return 1.0 * insolation;
}

function insolation(x, y, z) {
   var surfaceInsolation = 1.0;
   var insolation = surfaceInsolation - 0.5 * z;
   return Math.max(0.0, insolation);
}

function currentDirection(x, y, z) {
   if (x < 2 && y >= 2) {
      return {x: 0, y: -2, z: 0};
   } else if (y < 2 && x <= oceanSize().x - 3) {
      return {x: 2, y: 0, z: 0};
   } else if (x > oceanSize().x - 3 && y <= oceanSize().y - 3) {
      return {x: 0, y: 2, z: 0};
   } else if (y > oceanSize().y - 3) {
      return {x: -2, y: 0, z: 0};
   }
   return {x: 0, y: 0, z: 0};
}

function boundaryConditions() {
//   return "fixed";
   return "periodic";
//   return "mixed";
}
